cpp/validate_harvested_records.cc

```diff
@@ -322,126 +245,123 @@ void SendEmail(const std::string &email_address, const std::string &message_subj
-                  + ", field_name: \"" + field_name + "\"");
+void CheckGenericRequirements(const MARC::Record &record, std::string * const reasons_for_being_invalid) {
+    if (not record.hasTag("001"))
+        reasons_for_being_invalid->append("control field 001 is missing\n");
+    if (not record.hasTag("003"))
+        reasons_for_being_invalid->append("control field 003 is missing\n");
+    if (not record.hasTag("007"))
+        reasons_for_being_invalid->append("control field 007 is missing\n");
+
+    const auto _245_field(record.findTag("245"));
+    if (_245_field == record.end())
+        reasons_for_being_invalid->append("field 245 is missing\n");
+    else if (_245_field->getFirstSubfieldWithCode('a').empty())
+        reasons_for_being_invalid->append("subfield 245$a is missing\n");
+
+    if (not record.hasTag("100") and not record.hasTag("700"))
+        reasons_for_being_invalid->append("neither a field 100 nor 700 are present\n");
 }
 
 
-bool IsRecordValid(DbConnection * const db_connection, const MARC::Record &record, const GeneralInfo &general_info,
-                   std::map<std::string, JournalInfo> * const journal_name_to_info_map,
-                   unsigned * const new_record_count, std::string * const error_message)
+bool RecordIsValid(const MARC::Record &record, const std::vector<const FieldValidator *> &regular_article_field_validators,
+                   const std::vector<const FieldValidator *> &review_article_field_validators,
+                   std::string * const reasons_for_being_invalid)
 {
-    const std::string zeder_id(record.getFirstSubfieldValue("ZID", 'a'));
-    const std::string zeder_instance(record.getFirstSubfieldValue("ZID", 'b'));
-    if (zeder_id.empty() or zeder_instance.empty())
-        LOG_ERROR("Record w/ control number \"" + record.getControlNumber() + "\" has either no zeder_id or no zeder_instance!");
-
-    const auto journal_name(record.getSuperiorTitle());
-    if (journal_name.empty()) {
-        *error_message = "Record w/ control number \"" + record.getControlNumber() + "\" is missing a superior title!";
-        return false;
-    }
-
-    auto journal_name_and_info(journal_name_to_info_map->find(journal_name));
-    bool first_record(false); // True if the current record is the first encounter of a journal
-    if (journal_name_and_info == journal_name_to_info_map->end()) {
-        first_record = true;
-        JournalInfo new_journal_info;
-        LoadFromDatabaseOrCreateFromScratch(db_connection, zeder_id, zeder_instance, &new_journal_info);
-        (*journal_name_to_info_map)[journal_name] = new_journal_info;
-        journal_name_and_info = journal_name_to_info_map->find(journal_name);
+    reasons_for_being_invalid->clear();
+
+    const auto zid_field(record.findTag("ZID"));
+    if (unlikely(zid_field == record.end()))
+        LOG_ERROR("record is missing a ZID field!");
+    else if (zid_field->getFirstSubfieldWithCode('a').empty())
```

Comment:
Is it better to use a variable for the subfield lookup and re-using it instead of executing the lookup twice?

Reply:
Done